cupy/core/core.pyx

```diff
@@ -1723,6 +1721,7 @@ cpdef strides_t _get_strides_for_order_K(ndarray x, dtype, shape=None):
 
 
 _HANDLED_TYPES = (ndarray, numpy.ndarray)
+_KNOWN_TYPES = (ndarray, numpy.generic, bool, int, float, complex)
```

Comment:
Would suggest using [`numbers.Number`]( https://docs.python.org/3/library/numbers.html ) for these.

```python
In [1]: import numbers

In [2]: isinstance(True, numbers.Number)
Out[2]: True

In [3]: isinstance(1, numbers.Number)
Out[3]: True

In [4]: isinstance(1.5, numbers.Number)
Out[4]: True

In [5]: isinstance(1.5 + 1j, numbers.Number)
```

```suggestion
_KNOWN_TYPES = (ndarray, numpy.generic, numbers.Number)
```

Reply:
I think we can also drop `numpy.generic`, but will defer to @seberg

```python
In [1]: import numbers

In [2]: import numpy as np

In [3]: isinstance(np.float32(1), np.generic)
Out[3]: True

In [4]: isinstance(np.float32(1), numbers.Number)
Out[4]: True
```

```suggestion
_KNOWN_TYPES = (ndarray, numbers.Number)
```

Reply:
I am not sure any of these actually matter, since they all do not implement `__array_ufunc__` anyway.  One trick numpy does is to skip the `hasattr` check on builtin python types as a micro-optimization. I am not sure how much that actually helps.

Reply:
Right I think that's what is happening here as well. Also don't know how much it help

Reply:
Fair point, in which case you should swap the check and put `hasattr` second?  If this is about overhead/performance, I am not sure that `numbers.Number` is all that great though, I do not think `isinstance` checks are any faster than `hasattr` for ABCs.

Reply:
The overhead seems small ignorable, but I agree with your refactoring suggestions!